packages/@lwc/engine-core/src/framework/freeze-template.ts

```diff
@@ -147,84 +178,91 @@ function traverseStylesheets(
-                };
-            },
+        // this shim should be applied in both dev and prod
+        addLegacyStylesheetTokensShim(tmpl);
 
-            set(value) {
-                // If the value is null or some other exotic object, you would be broken anyway in the past
-                // because the engine would try to access hostAttribute/shadowAttribute, which would throw an error.
-                // However it may be undefined in newer versions of LWC, so we need to guard against that case.
-                this.stylesheetToken = isUndefined(value)
-                    ? undefined
-                    : (value as any).shadowAttribute;
-            },
-        });
-
-        // When ENABLE_FROZEN_TEMPLATE is false, warn in dev mode whenever someone is mutating the template
+        // When ENABLE_FROZEN_TEMPLATE is false, we want to warn in dev mode whenever someone is mutating the template
         if (process.env.NODE_ENV !== 'production') {
-            if (!isUndefined(tmpl.stylesheets)) {
-                warnOnStylesheetsMutation(tmpl.stylesheets);
-            }
-            for (const prop of TEMPLATE_PROPS) {
-                let value = tmpl[prop];
-                defineProperty(tmpl, prop, {
-                    enumerable: true,
-                    configurable: true,
-                    get() {
-                        return value;
-                    },
-                    set(newValue) {
-                        if (!mutationWarningsSilenced) {
-                            logError(
-                                `Dynamically setting the "${prop}" property on a template function ` +
-                                    `is deprecated and may be removed in a future version of LWC.`
-                            );
-                        }
-                        value = newValue;
-                    },
-                });
-            }
-
-            const originalDescriptor = getOwnPropertyDescriptor(tmpl, 'stylesheetTokens');
-            defineProperty(tmpl, 'stylesheetTokens', {
-                enumerable: true,
-                configurable: true,
-                get: originalDescriptor!.get,
-                set(value) {
-                    logError(
-                        `Dynamically setting the "stylesheetTokens" property on a template function ` +
-                            `is deprecated and may be removed in a future version of LWC.`
-                    );
-                    // Avoid logging twice (for both stylesheetToken and stylesheetTokens)
-                    mutationWarningsSilenced = true;
-                    originalDescriptor!.set!.call(this, value);
-                    mutationWarningsSilenced = false;
-                },
+            trackMutations(tmpl);
+        } else {
+            // In prod mode, we only track mutations if reporting is enabled
+            onReportingEnabled(() => {
+                trackMutations(tmpl);
```

Comment:
I'm a bit concerned that we will be doing all of this patching in prod mode, but we have been doing it in dev mode for a while and nobody complained.

In any case, it will have time to bake, and ideally we will have a gate to turn off reporting entirely, which will disable all mutation tracking.

Reply:
At this point, would it be worth having the ability to gate reporting by ReportEventId instead of all-or-nothing? As the number of patches we add grows, that might help us better manage potential breakages and unknown perf impacts.

Reply:
I think for now, all-or-nothing is definitely the simplest approach. Adding individual gates is always a pain; we will have enough headaches with just one.

Plus I think each one would probably need its own runtime flag in LWC. Whereas right now, it's just a simple matter of either using the `__unstable__ReportingControl` or not – if you don't use it, then the detection code is treeshaken out.

Once the reporting system proves itself as viable, I think we can look into enabling it piecemeal.